login/logout

8thDay opened the pull request:
Adds a `user` object to state in the App component.
Adds a method for updating the user object in state to the App component.
Adds a `Logout` component
App passes the new method to the UserForm and Logout components, so that they can update the user state.

UserForm returns both the JWT _and_ some user data. (id, username, image and imgURL).
UserForm stores the returned user data in state, then forwards the user to the dashboard.
Logout clears the user data from state, then forwards the user to the homepage.
The Google OAuth route currently return the user data and the JWT but **does not** update the user object in state or forward the user to a new page.

App passes part of the user object (currently the username) from state to the Header component. This data is used to determine if the user object in state is currently populated.
- if it is populated => a user is logged in
- if it isn't => no user is logged in
- The Header renders different menu options depending on whether the user is logged in/out

**Note** This code does not yet validate the JWT, and does not yet protect any routes.

8thDay committed:
Add user object to state in App component

Use presence or absence of username in user object
to determine which Header menu items are shown.

8thDay committed:
Return user object from login and signup routes

Google OAuth route also returns user object.
All login and signup routes return JWT.

8thDay committed:
Add user object to state in App component

Log in writes user data to state
Log out clears user data from state
User is forwarded to / on logout
User is forwarded to /dashboard on login
Google OAuth route not currently writing
user object to state or forwarding.

GarrusNapp reviewed (commented):
This looks cool! 
I have some suggestions: 

1. updateUserInState could receive just one object instead of 4 parameters, this would make code slightly easier to read in userform component
2. logged in/logged out navlinks could be put into their subcomponents, that'd make header more readable because of shorter arguments for ternary operator
3. I think that logout component doesn't need its own updateUser method, why not just call the one that's being passed in props?

8thDay committed:
Pass complete user object to updateUserInState()

Pass whole object to function, not separate params
Call updateUser method directly from props in
logOut component

8thDay committed:
Remove unused constructor from logout component

8thDay committed:
Divide header menu into separate subcomponents

8thDay commented:
@GarrusNapp Great suggestions, thank you. I've made the changes, and as you say, the code is much more readable. 👍 

I'm not sure why I was wrapping the updateUser method from props in another method... must have been a workaround for some syntax problem I was having! ¯\_(ツ)_/¯ 
Anyway, all fixed now!

8thDay committed:
Simplify user state in App component
